src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                    context.RegisterOperationBlockEndAction(blockAnalyzer.AnalyzeOperationBlockEnd);
+
+                    return;
+
+                    // Local Functions.
+                    bool HasSyntaxErrors()
+                    {
+                        foreach (var operationBlock in context.OperationBlocks)
+                        {
+                            if (operationBlock.SemanticModel.GetSyntaxDiagnostics(operationBlock.Syntax.Span, context.CancellationToken).HasAnyErrors())
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+                }
+
+                private void AnalyzeExpressionStatement(OperationAnalysisContext context)
+                {
+                    if (_options.UnusedValueExpressionStatementPreference == UnusedValuePreference.None)
+                    {
+                        return;
+                    }
+
+                    Debug.Assert(_options.UnusedValueExpressionStatementSeverity != ReportDiagnostic.Suppress);
+
+                    // Bail out cases for report unused expression value:
+                    //  1. Null type
+                    //  2. Void returning method invocations
+                    //  3. Bool returning method invocations
+                    //  4. Constant expressions
+                    //  5. Assignments, increment/decrement operations
+                    //  6. Invalid operation
+                    //  7. Expression and expression statement have differing first tokens, indicating explicit discard (for example, VB call statement)
+                    //  8. Special cases: Methods beloning to System.Threading.Interlocked and System.Collections.Immutable.ImmutableInterlocked.
+                    var expressionStatement = (IExpressionStatementOperation)context.Operation;
+                    var value = expressionStatement.Operation;
+                    if (value.Type == null ||
+                        value.Type.SpecialType == SpecialType.System_Void ||
+                        value.Type.SpecialType == SpecialType.System_Boolean ||
+                        value.ConstantValue.HasValue ||
+                        value is IAssignmentOperation ||
+                        value is IIncrementOrDecrementOperation ||
+                        value is IInvalidOperation ||
+                        value.Syntax.GetFirstToken() != expressionStatement.Syntax.GetFirstToken() ||
+                        (value is IInvocationOperation invocation &&
+                         (invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._interlockedType ||
+                          invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._immutableInterlockedType)))
```

Comment:
this last clause is such a pain to maintain.  I don't get the desire to jam everything into a massive expression that then has to be teased apart.  it would be totally fine to have that last part juts be another check that happens after these checks.